### ISSUE TITLE
Service generates empty twtxt feed for Echo JS Atom feed

### ISSUE THREAD

**Issue by vladimyr** [jointwt/rss2twtxt#10]
I'm not really a Golang guy so without deep diving into codebase here is what I know:
1. Submitting Echo JS homepage (https://echojs.com) results in an error because the service can't find the feed. This is actually the expected outcome because they fail to advertise it using rel links.
2. Submitting feed directly (https://echojs.com/rss) results in success but generated twtxt feed (https://feeds.twtxt.net/echo-js/twtxt.txt) is empty? :confused:

**Comment by prologic** [comment]
How long until you waited? The feed isn't fetched immediately but on the next cron cycle￼

**Comment by vladimyr** [comment]
Almost 24 hrs passed, I would guess cron executed at least a few times in the meantime?

BTW, I noticed that your posts often end in object replacement char. Is it my crappy phone or something odd is happening here? 🙃

![Screenshot_2021-01-25-08-29-30-201](https://user-images.githubusercontent.com/1170440/105674256-806cc380-5ee7-11eb-8c17-6ff0210c443b.jpeg)

**Comment by prologic** [comment]
Yeah I had a look and you're right. 👌 Time to debut! 🤣

Re obj char I have nfi what that's about?! 😂 I'm using the Firhub mobile app

**Comment by prologic** [comment]
I figured out why this particular RSS feed is turning up with an empty Twtxt feed.

See: https://github.com/jointwt/rss2twtxt/blob/04a3f6ba147327f279f87ad74bda4fe9238b6a2d/feeds.go#L162-L164

Basically the feed at https://echojs.com/rss is _actually_ invalid. None of the items have any `PublisheedDate`, which is kind of odd. I _could_ have sworn that this is a requirement field, but maybe it isn't? In any case, without a `PublishedDate`, its kind of pointless to try to generate a Twtxt feed out of it, as we kind of need a timestamp.

**Comment by vladimyr** [comment]
> Basically the feed at https://echojs.com/rss is _actually_ invalid. None of the items have any `PublisheedDate`, which is kind of odd. I _could_ have sworn that this is a requirement field, but maybe it isn't?

Actually, I went through the Atom format spec, and unfortunately, it isn't mandatory. The format is defined by [RFC4287](https://tools.ietf.org/html/rfc4287). There is a W3C run feed validator hosted at https://validator.w3.org/feed/ and their docs include a verbatim copy of [RFC4287](https://validator.w3.org/feed/docs/rfc4287.html#rfc.section.B) so it is safe to assume that service output is indeed correct and somewhat official.

FWIW, here is what it says about the Echo JS feed:  https://validator.w3.org/feed/check.cgi?url=https%3A%2F%2Fechojs.com%2Frss

> In any case, without a `PublishedDate`, its kind of pointless to try to generate a Twtxt feed out of it, as we kind of need a timestamp.

Maybe 🤔 I mean I understand your point but obviously, there is a spec mismatch here and whether or not it should be used as justification for rejecting some indeed valid fields is a hard question to answer.
Maybe we could use cron execution date to satisfy twtxt requirements? Using the same timestamp on multiple twts shouldn't be an issue right?

**Comment by prologic** [comment]
> Actually, I went through the Atom format spec, and unfortunately, it isn't mandatory. The format is defined by [RFC4287](https://tools.ietf.org/html/rfc4287). There is a W3C run feed validator hosted at https://validator.w3.org/feed/ and their docs include a verbatim copy of [RFC4287](https://validator.w3.org/feed/docs/rfc4287.html#rfc.section.B) so it is safe to assume that service output is indeed correct and somewhat official.

Damnit, I was afraid of that! 🤣 Also what a shame, 😳, who's dumb idea was it to make `PublishedDate` optional anyway 😂

> Maybe 🤔 I mean I understand your point but obviously, there is a spec mismatch here and whether or not it should be used as justification for rejecting some indeed valid feeds is a hard question to answer.
> Maybe we could use cron execution date to satisfy twtxt requirements? Using the same timestamp on multiple twts shouldn't be an issue right?

Yes we _could_ do this. I'd love your help on this one if you are willing? There is only one problem I can see here... How do we know if we've already seen a particular item? Right now the code uses the `PublishedDate` as a heuristic to know to skip over items it's seen before based on the `Last-Modified` date/time of the file it's producing.

Or in other words, you want to avoid duplicates. The service _could_ use the same [Twt Hash](https://dev.twtxt.net/doc/twthashextension.html) specification as what `twtd` pods use like [twtxt.net](https://twtxt.net) (_and others_), this would increase computational cost a bit, but that's _probably_ acceptable.

**Comment by vladimyr** [comment]
> Damnit, I was afraid of that! :rofl: Also what a shame, :flushed:, who's dumb idea was it to make `PublishedDate` optional anyway :joy:

That is an interesting design choice at least :upside_down_face: 

> Yes we _could_ do this. I'd love your help on this one if you are willing? 

A better question would be: Am I capable to help you with Golang implementation :smile: I'll give it a try.

> There is only one problem I can see here... How do we know if we've already seen a particular item? Right now the code uses the `PublishedDate` as a heuristic to know to skip over items it's seen before based on the `Last-Modified` date/time of the file it's producing.

Whenever questions like this arise I'm immediately inclined to check out any prior art. Let me go through, for example, `sfeed` sources and I'll get back on this one.

> Or in other words, you want to avoid duplicates. The service _could_ use the same [Twt Hash](https://dev.twtxt.net/doc/twthashextension.html) specification as what `twtd` pods use like [twtxt.net](https://twtxt.net) (_and others_), this would increase computational cost a bit, but that's _probably_ acceptable.

Yeah, that would be my answer too but let's see if there is any _cheaper_ option.

**Comment by vladimyr** [comment]
@prologic Seems like [`sfeed_update`](http://dezip.org/v1/8/https/codemadness.org/releases/sfeed/sfeed-0.8.tar.gz/sfeed-0.8/sfeed_update) just concatenates everything together and removes duplicates afterward. Uniqueness is checked by comparing `id`, `link`, and `title`. ([source](http://dezip.org/v1/8/https/codemadness.org/releases/sfeed/sfeed-0.8.tar.gz/sfeed-0.8/sfeed_update?search=merge#4))

`id` comes from either [`RSSTagGuid`](http://dezip.org/v1/8/https/codemadness.org/releases/sfeed/sfeed-0.8.tar.gz/sfeed-0.8/sfeed.c?search=RSSTagGuid#3) or [`AtomTagId`](http://dezip.org/v1/8/https/codemadness.org/releases/sfeed/sfeed-0.8.tar.gz/sfeed-0.8/sfeed.c?search=AtomTagId#3) i.e. there is already unique identifier present on every feed item. Now comes the funny part, in case of Echo JS that would be `guid` which is in fact the item URL itself :scream: 
```xml
<rss xmlns:atom="http://www.w3.org/2005/Atom" version="2.0">
<channel>
<title> Echo JS </title>
<link> http://www.echojs.com </link>
<description>Description pending</description>
<item>
<title> Intro to Choc UI, a Chakra UI Component library. </title>
<guid>https://dev.to/anubra266/intro-to-choc-ui-a-chakra-ui-component-library-5f1g</guid>
<link> https://dev.to/anubra266/intro-to-choc-ui-a-chakra-ui-component-library-5f1g </link>
<description>
<![CDATA[ <a href="http://www.echojs.com/news/38475">Comments</a> ]]>
</description>
<comments>http://www.echojs.com/news/38475</comments>
</item>
```

**Comment by prologic** [comment]
Yeah doing that kind of processing and rewriting the feed file every single time is not very efficient and some thing I'd rather avoid doing 🙃￼

**Comment by vladimyr** [comment]
The stupidest part of this saga is that Echo JS is actually opensource: https://github.com/echojs/echojs/blob/f15c5bc44e0b9c0d764f9f23256872f97da82af8/app.rb#L85-L102

And it is a perfect validation of shared wisdom: https://stackoverflow.com/a/28030380 To do it reliably you can't really avoid local caching:
> and you have to poll the whole feed and then extract the new ones (by diffing against caches on your end).

which is exactly what `sfeed_update` does. It is sad and inefficient but seems there is no way around it :weary:

**Comment by vladimyr** [comment]
With that being said there is a good piece of advice presented here: https://feedparser.readthedocs.io/en/latest/http-etag.html

>If you do not support ETag and Last-Modified headers, you will repeatedly download feeds that have not changed. This wastes your bandwidth and the publisher’s bandwidth, and the publisher may ban you from accessing their server.

Might easily happen with some sources.

**Comment by prologic** [comment]
I would opt for my later suggestion the of just using content addressing / hashing like we do for Twtxt pods already ￼

**Comment by prologic** [comment]
I will try and find some time to fix this over the next couple of days.￼

**Comment by vladimyr** [comment]
No pressure, after all, RSS still works :smile:

**Comment by prologic** [comment]
Maybe I can just use the GUID as a way to uniquely identify each item and keep a cache of this so that I can track the uniqueness of items?￼

**Comment by vladimyr** [comment]
Sure that will do it but you'll need to support both `<guid>` & `<id>` tags.

**Comment by prologic** [comment]
> Sure that will do it but you'll need to support both `<guid>` & `<id>` tags.

Can we rely on `guid` / `id` being mutually exclusive? Or _could_ they both exist and then we'd pick one over the otehr?

**Comment by vladimyr** [comment]
`sfeed` does so we could too :crossed_fingers:

**Comment by prologic** [comment]
Feel like learning a bit of Go? 😁
